freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/UmsUsersStateProvider.java

```diff
@@ -54,92 +77,233 @@
+                            umsUsersStateBuilder,
+                            usersStateBuilder,
+                            environmentIndex,
+                            userSyncStateModel,
+                            groups,
+                            wagNamesForOtherEnvironments);
+                    addServicePrinciplesCloudIdentities(
+                            accountId, environmentCrn, umsUsersStateBuilder, requestIdOptional);
+                    umsUsersStateBuilder.setUsersState(usersStateBuilder.build());
+                    umsUsersStateMap.put(environmentCrn, umsUsersStateBuilder.build());
+                });
+        return umsUsersStateMap;
+    }
 
-                users.forEach(u -> {
-                    FmsUser fmsUser = umsUserToUser(u);
-                    // add workload username for each user. This will be helpful in getting users from IPA.
-                    umsUsersStateBuilder.addRequestedWorkloadUsers(fmsUser);
+    @VisibleForTesting
+    Map<String, UmsUsersState> getUmsUsersStateMap(
+            String accountId, String actorCrn, Collection<String> environmentCrns,
+            Set<String> userCrns, Set<String> machineUserCrns, Optional<String> requestIdOptional,
+            boolean fullSync) {
+        List<User> users = getUsers(actorCrn, accountId, requestIdOptional, fullSync, userCrns);
+        List<MachineUser> machineUsers = getMachineUsers(actorCrn, accountId, requestIdOptional, fullSync, machineUserCrns);
+
+        Map<String, FmsGroup> crnToFmsGroup = convertGroupsToFmsGroups(
+                grpcUmsClient.listAllGroups(INTERNAL_ACTOR_CRN, accountId, requestIdOptional));
+        Map<WorkloadAdministrationGroup, FmsGroup> wags = convertWagsToFmsGroups(
+                grpcUmsClient.listWorkloadAdministrationGroups(INTERNAL_ACTOR_CRN, accountId, requestIdOptional));
+        List<String> requestedWorkloadUsernames = Streams.concat(
+                users.stream().map(User::getWorkloadUsername),
+                machineUsers.stream().map(MachineUser::getWorkloadUsername))
+                .collect(Collectors.toList());
+
+        Map<String, UmsUsersState> umsUsersStateMap = new HashMap<>();
+        environmentCrns.forEach(environmentCrn -> {
+            UmsUsersState.Builder umsUsersStateBuilder = new UmsUsersState.Builder()
+                    .setWorkloadAdministrationGroups(wags.values());
+            UsersState.Builder usersStateBuilder = new UsersState.Builder();
+            addRequestedWorkloadUsernames(umsUsersStateBuilder, requestedWorkloadUsernames);
+            addGroupsToUsersStateBuilder(usersStateBuilder, crnToFmsGroup.values());
+            Set<String> wagNamesForOtherEnvironments =
+                    addWagsToUsersStateBuilder(usersStateBuilder, wags, environmentCrn);
+
+            EnvironmentAccessChecker environmentAccessChecker = createEnvironmentAccessChecker(environmentCrn);
+            TriConsumer<String, FmsUser, List<CloudIdentity>> actorHandler =
```

Comment:
I have to say this class is getting more and more complex, the methods are getting more complex and bigger. It's pretty hard to understand and review it actually. I would like to kindly ask you to refactor this before/while putting these new bulk sync logic into it.
eg this `actorHandler` might be a separate class as it has a try-catch, which invokes a method with 10 parameters, 3 of them are lambdas

Reply:
refactored.
I pulled out the common functionality between the bulk and default implementations into an ActorHandler and this TriConsumer to a separate method